binary_search_tree/tree.py

```diff
@@ -7,49 +10,135 @@ def __init__(self, key, val = None):
+            else:
+                current_node.left = new_node
+                return
+        else:
+            if current_node.right:
+                return self.add_helper(current_node.right, new_node)
+            else:
+                current_node.right = new_node
+                return
+
+    # Time Complexity: O(log n)
+    # Space Complexity: O(1)
     def add(self, key, value = None):
-        pass
+        new_node = TreeNode(key, value)
+        if self.root == None:
+            self.root = new_node
+        else:
+            self.add_helper(self.root, new_node)
+
+    def find_helper(self, node, key):
+        if node == None:
+            return None
+        if node.key == key:
+            return node.value
+        if key < node.key:
+            return self.find_helper(node.left, key)
+        else:
+            return self.find_helper(node.right, key)
 
-    # Time Complexity: 
-    # Space Complexity: 
+    # Time Complexity: O(log n)
+    # Space Complexity: O(1)
     def find(self, key):
-        pass
+        return self.find_helper(self.root, key)
+
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)
+    def inorder_helper(self, node, arr):
+        if node == None:
+            return
+        self.inorder_helper(node.left, arr)
+        arr.append({"key": node.key, "value": node.value})
+        self.inorder_helper(node.right, arr)
 
-    # Time Complexity: 
-    # Space Complexity: 
     def inorder(self):
-        pass
+        arr = []
+
+        self.inorder_helper(self.root, arr)
+        return arr
+
+    def preorder_helper(self, node, arr):
+        if node == None:
+            return
+        
+        arr.append({"key": node.key, "value": node.value})
+        self.preorder_helper(node.left, arr)
+        self.preorder_helper(node.right, arr)
 
-    # Time Complexity: 
-    # Space Complexity:     
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)  
     def preorder(self):
-        pass
+        arr = []
+
+        self.preorder_helper(self.root, arr)
+        return arr
 
-    # Time Complexity: 
-    # Space Complexity:     
+    def postorder_helper(self, node, arr):
+        if node == None:
+            return
+        
+        self.postorder_helper(node.left, arr)
+        self.postorder_helper(node.right, arr)
+        arr.append({"key": node.key, "value": node.value})
+
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)  
     def postorder(self):
-        pass
+        arr = []
+        self.postorder_helper(self.root, arr)
+        return arr
 
-    # Time Complexity: 
-    # Space Complexity:     
-    def height(self):
-        pass
+    def height_helper(self, node):
+        if not node:
+            return 0
+        else:
+            return 1 + max(self.height_helper(node.left), self.height_helper(node.right))
 
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)  
+    def height(self):
```

Comment:
👍 However the space complexity is O(n) for an unbalanced tree or O(log n) for a balanced tree.